qa/rpc-tests/orchard_reorg.py

```diff
@@ -0,0 +1,146 @@
+
+    def run_test(self):
+        # Activate NU5 so we can test Orchard.
+        self.nodes[0].generate(10)
+        self.sync_all()
+
+        # Generate a UA with only an Orchard receiver.
+        account = self.nodes[0].z_getnewaccount()['account']
+        addr = self.nodes[0].z_getaddressforaccount(account, ['orchard'])
+        assert_equal(addr['account'], account)
+        assert_equal(set(addr['pools']), set(['orchard']))
+        ua = addr['unifiedaddress']
+
+        # Before mining any Orchard notes, finalorchardroot should be the empty Orchard root.
+        assert_equal(
+            ORCHARD_TREE_EMPTY_ROOT,
+            self.nodes[0].getblock(self.nodes[0].getbestblockhash())['finalorchardroot'],
+        )
+
+        # finalorchardroot should not change if we mine additional blocks without Orchard notes.
+        self.nodes[0].generate(100)
+        self.sync_all()
+        assert_equal(
+            ORCHARD_TREE_EMPTY_ROOT,
+            self.nodes[0].getblock(self.nodes[0].getbestblockhash())['finalorchardroot'],
+        )
+
+        # Create an Orchard note.
+        recipients = [{'address': ua, 'amount': Decimal('12.5')}]
+        opid = self.nodes[0].z_sendmany(get_coinbase_address(self.nodes[0]), recipients, 1, 0)
+        wait_and_assert_operationid_status(self.nodes[0], opid)
+
+        # After mining a block, finalorchardroot should have changed.
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+        orchardroot_oneleaf = self.nodes[0].getblock(self.nodes[0].getbestblockhash())['finalorchardroot']
+        print("Root of Orchard commitment tree with one leaf:", orchardroot_oneleaf)
+        assert(orchardroot_oneleaf != ORCHARD_TREE_EMPTY_ROOT)
+
+        # finalorchardroot should not change if we mine additional blocks without Orchard notes.
+        self.nodes[0].generate(4)
+        self.sync_all()
+        assert_equal(
+            orchardroot_oneleaf,
+            self.nodes[0].getblock(self.nodes[0].getbestblockhash())['finalorchardroot'],
+        )
+
+        # Split the network so we can test the effect of a reorg.
+        print("Splitting the network")
+        self.split_network()
+
+        # Create another Orchard note on node 0.
+        recipients = [{'address': ua, 'amount': Decimal('12.5')}]
+        opid = self.nodes[0].z_sendmany(get_coinbase_address(self.nodes[0]), recipients, 1, 0)
+        wait_and_assert_operationid_status(self.nodes[0], opid)
+
+        # Mine two blocks on node 0.
+        print("Mining 2 blocks on node 0")
+        self.nodes[0].generate(2)
+        self.sync_all()
+        orchardroot_twoleaf = self.nodes[0].getblock(self.nodes[0].getbestblockhash())['finalorchardroot']
+        print("Root of Orchard commitment tree with two leaves:", orchardroot_twoleaf)
+        assert(orchardroot_twoleaf != ORCHARD_TREE_EMPTY_ROOT)
+        assert(orchardroot_twoleaf != orchardroot_oneleaf)
+
+        # Generate 2 blocks on node 2.
```

Comment:
```suggestion
        # Generate 10 blocks on node 2.
```
However, I think that 10 blocks on node 2 are not guaranteed to outweigh 2 blocks on chain 0 (even though it's unlikely that they won't). Perhaps we should generate until the total work of the node 2 chain exceeds that of the node 0 chain.

Reply:
Oh, work is [defined](https://zips.z.cash/protocol/protocol.pdf#workdef) by the difficulty target rather than the actual hash, so this will always operate as intended.